### ISSUE TITLE
TypeError: Cannot read property 'composed' of undefined

### ISSUE THREAD

**Issue by mbulfair** [lerna/lerna#1897]
_For security reasons, I can't share some debug logs with sensitive material. So areas that have I used `REMOVED`_  

## Expected Behavior

Running `lerna publish` should prompt me for all changes since last release

## Current Behavior

Running
`npx lerna publish --git-remote upstream --conventional-commits --no-verify-registry --no-verify-access --registry=REMOVED`

```
5 error TypeError: Cannot read property 'composed' of undefined
5 error     at new Command (/Users/REMOVED/Projects/REMOVED/components-react/node_modules/@lerna/command/index.js:30:33)
5 error     at new VersionCommand (/Users/REMOVED/Projects/REMOVED/components-react/node_modules/@lerna/version/index.js:40:1)
5 error     at factory (/Users/REMOVED/Projects/REMOVED/components-react/node_modules/@lerna/version/index.js:37:10)
```
<details><summary>lerna.json</summary><p>

```json
{
  "command": {
    "publish": {
      "message": "chore: publish"
    }
  },
  "packages": [
    "components/*",
    "tools/*"
  ],
  "version": "independent"
}
```
</p></details>

## Context

We use a private NPM solution Nexus Repository Manager, which is not 100% a true NPM solution, we are locked to the version of lerna that still uses the no-registry flag. Unless there's a way around this. With that version, when we run the lerna publish and it fails, it's very difficult to get it back into a place that it knows there are changes again. It seems lerna thinks it's published, but it never got to our private server. Any help with this would be appreciated. 

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.3.2 |
| `npm --version`  | 6.4.1 |
| `node --version` | 10.15.0 |

| OS | Version |
| --- | --- |
| NAME | VERSION |
| macOS Mojave | 10.14.4 |

**Comment by evocateur** [comment]
Since [v3.8.3](https://github.com/lerna/lerna/releases/tag/v3.8.3) lerna publish no longer attempts to validate registry auth for any registry that isn't literally `"https://registry.npmjs.org/"` (https://github.com/lerna/lerna/commit/b44f2f9). I would try updating to the latest. The error itself is rather odd, since I run a very similar command whenever I publish Lerna itself...

**Comment by mbulfair** [comment]
@evocateur I had it at the latest version, 3.10.3 I think, and I would see results of errors with 'basic' authentication. Because we had passed that --no-verify-registry flag. It looked to be a 404 error with the latest changes.

**Comment by mbulfair** [comment]
`lerna http fetch PUT 401 REMOVED(PATH to NEXUS) 409ms
lerna ERR! E401 Unable to authenticate, need: BASIC realm="Sonatype Nexus Repository Manager"` @evocateur

**Comment by evocateur** [comment]
Really wish JFrog/Artifactory/Nexus would get on board the bearer token auth train, legacy auth is extremely annoying...

**Comment by mbulfair** [comment]
@evocateur seems so, it might also be the way we are authenticating with it.  This is with Nexus, we also have Artifactory, i'm not sure if they will be the same.

**Comment by mbulfair** [comment]
Where does lerna publish get the credentials to do a npm publish for each component? in our setup we have a `.npmrc` which has a `_auth` and a exported encoded value. @evocateur

**Comment by mbulfair** [comment]
I managed to get it to publish with using Artifactory, which sent the proper responses to stop registration check, I will say the npm publish failed because it needed the _auth in each package repo, with a .npmrc file

**Comment by thomas-darling** [comment]
@evocateur
We're getting the same authentication error with `Azure Artifacts`, i.e. `<company>.pkgs.visualstudio.com`.

While I fully appreciate the frustration caused by the private registries not keeping up, or having quirks, I hope you can appreciate the frustration it causes us, when we can't publish our packages. Something needs to be done about this, or we'll have to abandon Lerna - it's a really awesome tool, but right now, this is simply broken for us.

From my perspective, since we _can_ successfully publish by running `npm publish` in each package, `lerna publish` should work too - if it doesn't, then that's a bug in Lerna, not in the registry.

---

In `package.json` in each package:
```
"publishConfig": {
  "registry": "https://<hidden>.pkgs.visualstudio.com/_packaging/<hidden>/npm/registry/",
  "verifyAccess": false
},
```

In `.npmrc`, located in each package:
```
registry=https://<hidden>.pkgs.visualstudio.com/_packaging/<hidden>/npm/registry/
always-auth=true
; begin auth token
//<hidden>.pkgs.visualstudio.com/_packaging/<hidden>/npm/registry/:_authToken=<hidden>
//<hidden>.pkgs.visualstudio.com/_packaging/<hidden>/npm/:_authToken=<hidden>
; end auth token
```

Output when running `lerna publish`:
```
...
lerna info publish Publishing packages to npm...
lerna info Verifying npm credentials
lerna http fetch GET 200 https://registry.npmjs.org/-/npm/v1/user 1297ms
lerna http fetch GET 200 https://registry.npmjs.org/-/org/thomas-darling/package?format=cli 246ms
lerna http fetch PUT 401 https://<hidden>.pkgs.visualstudio.com/_packaging/<hidden>/npm/registry/<hidden> 439ms
lerna ERR! E401 Unable to authenticate, need: Bearer authorization_uri=https://login.windows.net/<hidden>, Basic realm="https://pkgsprodsu3weu.app.pkgs.visualstudio.com/", TFS-Federated
```

Also, why is it even hitting `registry.npmjs.org` here?

**Comment by mbulfair** [comment]
@thomas-darling taking a guess, but did this work in an earlier version? Also when you run the `lerna publish` try to pass the registry there, in my setup I don't have it in each package. 

Lastly I do have in each of the components `package.json` based on your log verus my log, the first two lines about `https://registry.npmjs.org/` looks like it's not reading the local pacakge.json values.

```
  "publishConfig": {
    "access": "public"
  },
```
All of the <hidden> are the private server.
```
lerna info publish Publishing packages to npm...
lerna http fetch GET 404 https://<hidden>/-/org/anonymous/package?format=cli 190ms
lerna http fetch GET 404 https://<hidden>/-/user/anonymous/package?format=cli 60ms
lerna WARN EREGISTRY Registry "<hidden>" does not support `npm access ls-packages`, skipping permission checks...
```

**Comment by thomas-darling** [comment]
@mbulfair It's actually the first time I'm using Lerna with Azure, so can't say - and of course, it's entirely possible that I'm just doing something wrong here, but if that's the case, I can't figure out what :-)

You are right about the `registry` in the `package.json` files being ignored though - so that would appear to be a bug, as it should work according to the [documentation](https://github.com/lerna/lerna/tree/master/commands/publish#publishconfigregistry). Or am I missing something, @evocateur?

If I pass the registry as a command line argument, I get this instead - still the same error though:

```
lerna info publish Publishing packages to npm...
lerna notice Skipping all user and access validation due to third-party registry
lerna notice Make sure you're authenticated properly ¯\_(ツ)_/¯
lerna WARN ENOLICENSE Packages <hidden>/infrastructure are missing a license
lerna http fetch PUT 401 https://<hidden>.pkgs.visualstudio.com/_packaging/<hidden>/npm/registry/<hidden> 389ms
lerna ERR! E401 Unable to authenticate, need: Bearer authorization_uri=https://login.windows.net/<hidden>, Basic realm="https://pkgsprodsu3weu.app.pkgs.visualstudio.com/", TFS-Federated
```

Also, this is really not something I want to set as a command line argument. Things like this really need to be configured somewhere, so there's no risk of accidentally publishing to the wrong place - even just leaking the _existence_ of a package could be a problem, as that could reveal what a company is working on. And why do we even need this, when it's already in the `.npmrc` file?

Regarding the `"access": "public"` - I'm not really sure that's something we want to set. We most certainly do not want those packages to be publicly available, so I'm not really sure I want to even test that - especially when I don't feel certain about which registry it's hitting. Also, it wouldn't make much sense if that was the problem here, based on the error message...

**Comment by evocateur** [comment]
`"access": "public"` is only necessary for scoped packages (`@my-scope/pkg-name`) that are published to the public registry, which treats scoped packages as private by default. If your registry does not support this awareness, the `publishConfig.access` field is completely ignorable.

As for `npm publish` still working, what version _exactly_ of npm are we talking about? If it's `<=6.5.0`, it's still using the legacy registry client. [v6.6.0](https://github.com/npm/cli/blob/latest/CHANGELOG.md#v660-2019-01-17) replaced all of that with `libnpm`-based extractions, which [v6.7.0](https://github.com/npm/cli/blob/latest/CHANGELOG.md#v670-2019-01-23) includes a fix for [a bug involving legacy auth](https://npm.community/t/npm-6-6-0-broke-authentication-with-npm-registry-couchapp/4752).

Lerna has been using `libnpm` since [v3.7.0](https://github.com/lerna/lerna/releases/tag/v3.7.0) to publish instead of calling `npm publish` from a subprocess. If you use a lockfile, you may need to "refresh" lerna to pick up the patched version of [libnpmpublish](https://github.com/npm/libnpmpublish/commit/38422d0) and [npm-registry-fetch](https://github.com/npm/npm-registry-fetch/commit/a91f90c).

```
npm rm -D lerna && npm i -D lerna@latest
```

The original error described in this issue would likely happen if the installation of lerna was somehow nested, perhaps during a migration to a new version (post [v3.10.0](https://github.com/lerna/lerna/releases/tag/v3.10.0)) that specifies sibling dependencies with exact version ranges (no `^`). But that's a weird edge-case.

**Comment by mbulfair** [comment]
So turns out, on friday when I did a clean install, it installed 3.6.0, I forced to update to 3.10.8 and then I noticed lerna was just skipping everything and giving me a error. I noticed I wasn't logged in. I thought having a .npmrc file in each component was all I needed. Seems to get it to publish, I have to move all of that to the machine level .npmrc, log in first, and then publish works. Is that how it's expected to work?
